### PR TITLE
498 circuit size estimation

### LINKED ISSUE #498: Special assigner mode for circuit size estimation
The users want to have a way of estimating the proving time without actually running the prover. The circuit size calculation seems to be the best way.

Sometimes (as in the case of bls12-381), we do not even have the prover, so the circuit size is actually `the only way` to estimate how big the circuit is.

We need to add a special assigner mode in which different stats of the result circuit will be printed.

Circuit stats to count:
- rows_amount
- gates_amount

If you understand that we can provide any additional params to estimate the quality of the circuit and proving time - add them to the list above.

As I see it, it makes sense to calculate some (or maybe all) of the circuit stats without actually building the circuit. For example, we can use component's `rows_amount` instead of actually building the assignment table.

### LINKED ISSUE #527: [Possible bug] Nested "if" fails assertion "(stack_size - 1) == call_stack.size()"
Example 
`using namespace nil::crypto3::algebra::curves;

[[circuit]] typename pallas::base_field_type::value_type
    field_arithmetic_example(typename pallas::base_field_type::value_type a,
                             typename pallas::base_field_type::value_type b) {

        if (a <= b) {
            if (a == b) {
                return a + b;
            }
            else {
                return a - b;
            }
        }
        if (a >= b) {
            if (a > b) {
                return a * b;
            }
            else {
                return a / b;
            }
        }

}`

### PR DESCRIPTION
**Pull request by CblPOK-git**
Closes #498 
Closes https://github.com/NilFoundation/zkllvm-blueprint/issues/320
Closes #527

### CODE REVIEW

File: bin/assigner/src/main.cpp [hunk 1]
```diff
@@ -661,20 +668,34 @@ int main(int argc, char *argv[]) {
...
-        std::cerr << "Invalid command line argument - assignment table file name is not specified" << std::endl;
-        std::cout << options_desc << std::endl;
-        return 1;
-    }
-
-    if (vm.count("circuit")) {
-        circuit_file_name = vm["circuit"].as<std::string>();
+    if (vm.count("estimate_size")) {
+        if (vm.count("assignment-table")) {
+            std::cerr << "Assigner is running with --estimate_size flag and will not generate .tbl assignment table.\n";
```
Comment (akokoshn) [inline bin/assigner/src/main.cpp:673]: I guess better allow user set output files (*.tbl, *.crct) even set `--estimate-size` flag. It's not error, may be just warning.
```
if (vm.count("assignment-table")) {
        assignment_table_file_name = vm["assignment-table"].as<std::string>();
    } else if (!vm.count("estimate_size")) {
        std::cerr << "Invalid command line argument - assignment table file name is not specified" << std::endl;
        std::cout << options_desc << std::endl;
        return 1;
    }
```

File: examples/CMakeLists.txt [hunk 2]
```diff
@@ -49,6 +49,15 @@ function(assign_ir target binary_name input private_input curve_type max_num_pro
         COMMAND_EXPAND_LISTS
         VERBATIM)
 
+    add_custom_target(${target}_estimate_size
```
Comment (makxenov) [inline examples/CMakeLists.txt:52]: How about some tests on this mode? Even simple runs of examples without checking results would be better then nothing.

Comment (CblPOK-git) [reply examples/CMakeLists.txt:52]: done

**Review by akokoshn — APPROVED** [review]

**Comment by aleasims** [comment]
This PR (changes in assigner) brings a bunch of new warnings on compilation, which seem to be simply fixable:

```plain
zkllvm/libs/assigner/include/nil/blueprint/component_mockups/comparison.hpp:315:29: warning: enumeration value 'FLAG' not handled in switch [-Wswitch]
...
```

In order to keep our code clean probably it's better to fix this right away and don't wait until it will result in strange error.

**Pull request merged**